### ISSUE TITLE
Unable to join versions association

### ISSUE THREAD

**Issue by seanlinsley** [paper-trail-gem/paper_trail#1135]
When upgrading Paper Trail from 8.1.2 to 0e94b68c232253b28178e73d129157a2f68e2954, we started running into a Rails error anywhere we performed a join:

```ruby
>> User.joins(:versions).to_a

ArgumentError: The association scope 'versions' is instance dependent (the scope block takes an argument). 
Preloading instance dependent scopes is not supported.

from gems/activerecord-5.1.5/lib/active_record/reflection.rb:481:in `check_preloadable!'
```

What inside Paper Trail caused this error? Is it something we'll have to upgrade Rails to fix?

**Comment by jaredbeck** [comment]
hmm I don't know. Can you try taking smaller steps? eg. upgrade from 8.1.2 to 9.0.0, etc.?

**Comment by seanlinsley** [comment]
ac710b3197a0ed2cf493b1229ab31ca4c70e5c9c doesn't produce this error, but the next commit, 58369e1d8fe16ae7f37dee3da05c120827afdf3e from #1108 does.

**Comment by seanlinsley** [comment]
From the error message, it appears that the issue is that the association is set up with a block that accepts an argument:

https://github.com/paper-trail-gem/paper_trail/blob/2e697fd4777fbd759bc93825c68cc089a3773e4d/lib/paper_trail/model_config.rb#L141-L144

**Comment by seanlinsley** [comment]
Are we okay with going back to the old association definition for non-STI models, with the understanding that STI models won't be able to join the versions association without custom SQL?

**Comment by seanlinsley** [comment]
It doesn't look like there's a test covering `joins(:version)`, which would explain why y'all didn't catch it during that PR. I'll go ahead and start working on a PR.

**Comment by seanlinsley** [comment]
I found this method in the Rails docs: https://apidock.com/rails/v4.2.7/ActiveRecord/Inheritance/ClassMethods/descends_from_active_record%3F

>Returns true if this does not need STI type condition. Returns false if STI type condition needs to be applied.

Which is close, but wouldn't we want the same behavior for the base class?

```ruby
> ::Family::CelebrityFamily.descends_from_active_record?
=> false
> ::Family::Family.descends_from_active_record?
=> true
```

We might just want to use this (copied from the method's source):
```ruby
columns_hash.include?(inheritance_column)
```

**Comment by seanlinsley** [comment]
Turns out I couldn't use `descends_from_active_record?` b/c in the family test case, Paper Trail is configured on the base class & then it's expected to be available on every child class.

I'd do this instead, but it causes an error b/c the database isn't available at this point in time.
```ruby
    def setup_versions_association(klass)
      if klass.columns_hash.exclude?(klass.inheritance_column)
        @model_class.has_many(
          @model_class.versions_association_name,
          -> { order(model.timestamp_sort_order) },
          class_name: @model_class.version_class_name,
          as: :item
        )
      else
        klass.has_many(
          klass.versions_association_name,
          lambda do |object|
```

**Comment by seanlinsley** [comment]
Another option would be to rely on the [`inherited` callback](http://ruby-doc.org/core-2.5.1/Class.html#method-i-inherited), redefining the association on every STI child class that's defined.

**Comment by seanlinsley** [comment]
Hmm never mind, `descends_from_active_record?` has the same issue b/c it also relies on `columns_hash`.

But if there's a way to know if it's an STI model without breaking the load order, we could do something like this to re-evaluate when models inherit from others:
```ruby
      klass.singleton_class.prepend Module.new {
        def inherited(klass)
          super
          paper_trail.send(:setup_versions_association, klass)
        end
      }
```

**Comment by seanlinsley** [comment]
Oh, actually the block is only accepting an argument in order to get access to the `paper_trail` object on the model. That's not needed if we redefine the association via an `inherited` callback; we can just rely on the `klass` local variable.

**Comment by seanlinsley** [comment]
Okay that seems to be working. The issue I'm now running into is that the `has_many` correctly uses the subclass, but we stored the base class in the database. Where do we set the item type? Oddly, only the destroy event specifies `item_type` and `item_id`.

**Comment by seanlinsley** [comment]
Looks like we might be relying on the Rails association to set them. From `record_create`:
```ruby
      versions_assoc = @record.send(@record.class.versions_association_name)
      version = versions_assoc.new(data)
```

I'm guessing the issue is that the association builder auto-generates `item_type`, while we're defining a block that overrides it when querying the database.

So the fix would be to manually specify the `item_type` in every event, like `destroy` currently does.

**Comment by jaredbeck** [comment]
Fixed by #1143
